### PR TITLE
AEA-935 skill to skill message send

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

skill to skill message send
## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/agent.py [hunk 1]
```diff
@@ -290,3 +295,37 @@ def exception_handler(
             f"Exception {repr(exception)} raised during {repr(function)} call."
         )
         return True
+
+
+class SkillOutBox(OutBox):
```
Comment (DavidMinarsch) [inline aea/agent.py:300]: The concept of skill does not exist on the `Agent` level, should be added on `AEA` level

File: aea/agent_loop.py [hunk 2]
```diff
@@ -171,6 +187,18 @@ def __init__(
         self._agent: AbstractAgent = self._agent
 
         self._periodic_tasks: Dict[Callable, PeriodicCaller] = {}
+        self._skill2skill_message_queue: Optional[asyncio.Queue] = None
+
+    def _setup(self) -> None:
```
Comment (DavidMinarsch) [inline aea/agent_loop.py:192]: Missing docstring

File: tests/common/utils.py [hunk 3]
```diff
@@ -305,6 +306,18 @@ def wait_for_condition(condition_checker, timeout=2, error_msg="Timeout", period
             raise TimeoutError(error_msg)
 
 
+async def wait_for_condition_async(
```
Comment (DavidMinarsch) [inline tests/common/utils.py:309]: I feel like we have many helpers like this...

File: tests/test_aea.py [hunk 4]
```diff
@@ -288,7 +288,7 @@ def test_handle():
                 message=encoded_msg,
             )
             # send envelope via localnode back to agent/bypass `outbox` put consistency checks
-            aea.outbox._multiplexer.put(envelope)
+            aea.runtime.multiplexer.put(envelope)
```
Comment (DavidMinarsch) [inline tests/test_aea.py:291]: see comment above

File: aea/agent.py [hunk 5]
```diff
@@ -273,7 +273,10 @@ def get_message_handlers(self) -> List[Tuple[Callable[[Any], None], Callable]]:
 
         :return: List of tuples of callables: handler and coroutine to get a message
         """
-        return [(self.handle_envelope, self.inbox.async_get)]
+        return [
+            (self.handle_envelope, self.inbox.async_get),
+            (self.handle_envelope, self.runtime.agent_loop.skill2skill_queue.get),
```
Comment (DavidMinarsch) [inline aea/agent.py:278]: Should be added on aea level in same method. Agent should not know about skill2skill_queue

File: aea/agent_loop.py [hunk 6]
```diff
@@ -171,6 +198,46 @@ def __init__(
...
+        if not self._skill2skill_message_queue:
+            raise ValueError("_skill2skill_message_queue is not set!")
+        return self._skill2skill_message_queue
+
+    def send_to_skill(
+        self,
+        message_or_envelope: Union[Message, Envelope],
+        context: Optional[EnvelopeContext] = None,
+    ) -> None:
+        """Send message or envelope to another skill."""
```
Comment (DavidMinarsch) [inline aea/agent_loop.py:219]: missing params in docstrings

File: aea/context/base.py [hunk 7]
```diff
@@ -87,6 +88,17 @@ def __init__(
...
         self._data_dir = data_dir
         self._namespace = SimpleNamespace(**kwargs)
+        self._send_to_skill = send_to_skill
+
+    def send_to_skill(
+        self,
+        message_or_envelope: Union[Message, Envelope],
+        context: Optional[EnvelopeContext] = None,
+    ) -> None:
+        """Send message or envelope to another skill."""
```
Comment (DavidMinarsch) [inline aea/context/base.py:98]: docstring

File: aea/skills/base.py [hunk 8]
```diff
@@ -252,6 +253,16 @@ def __getattr__(self, item: Any) -> Any:
         """Get attribute."""
         return super().__getattribute__(item)  # pragma: no cover
 
+    def send_to_skill(
+        self,
+        message_or_envelope: Union[Message, Envelope],
+        context: Optional[EnvelopeContext] = None,
+    ) -> None:
+        """Send message or envelope to another skill."""
```
Comment (DavidMinarsch) [inline aea/skills/base.py:261]: docstring incomplete

File: aea/agent_loop.py [hunk 9]
```diff
@@ -91,6 +93,18 @@ def state(self) -> AgentLoopStates:
         """Get current main loop state."""
         return self._state.get()
 
+    async def wait_state(
+        self, state_or_states: Union[Any, Sequence[Any]]
+    ) -> Tuple[Any, Any]:
+        """Wait state to be set.
```
Comment (DavidMinarsch) [inline aea/agent_loop.py:99]: newline

File: tests/test_aea.py [hunk 10]
```diff
@@ -898,3 +898,49 @@ def aea_action(self):
         self.aea_tool.aea.runtime.agent_loop._execution_control(
             self.behaviour.act_wrapper
         )
+
+
+def test_skill2skill_message():
+    """Tests message can be sent directly to any skill."""
```
Comment (DavidMinarsch) [inline tests/test_aea.py:904]: wrap in temporary directory

File: tests/common/utils.py (hunk 3 above)
Comment (solarw) [reply tests/common/utils.py:309]: I made sync one, not sure async is present

File: aea/agent_loop.py (hunk 9 above)
Comment (solarw) [reply aea/agent_loop.py]: done

File: tests/test_aea.py (hunk 10 above)
Comment (solarw) [reply tests/test_aea.py:905]: done

File: aea/skills/base.py (hunk 8 above)
Comment (solarw) [reply aea/skills/base.py]: done

File: aea/context/base.py (hunk 7 above)
Comment (solarw) [reply aea/context/base.py]: done

File: aea/agent_loop.py (hunk 6 above)
Comment (solarw) [reply aea/agent_loop.py]: done

File: aea/agent.py (hunk 5 above)
Comment (solarw) [reply aea/agent.py]: done

**Review by DavidMinarsch — APPROVED** [review]
LGTM

File: aea/agent_loop.py [hunk 11]
```diff
@@ -166,11 +200,58 @@ def __init__(
...
+
+    def _setup(self) -> None:
+        """Set up agent loop before started."""
+        self._skill2skill_message_queue = asyncio.Queue()
+        super()._setup()
+
+    @property
+    def skill2skill_queue(self) -> Queue:
+        """Get skill to skill message queue."""
+        if not self._skill2skill_message_queue:  # pragma: nocover
```
Comment (marcofavorito) [inline aea/agent_loop.py:219]: does this work, even if the queue is empty?

Comment (marcofavorito) [reply aea/agent_loop.py:219]: what I mean is: the condition should evaluate to true only if it is `None`, and not if it is empty

**Review by marcofavorito — APPROVED** [review]
LGTM

File: aea/agent_loop.py (hunk 11 above)
Comment (solarw) [reply aea/agent_loop.py:219]: it works ok, queue has special method 'empty' for this

**Pull request merged**